lib/src/main/java/com/knuddels/jtokkit/TokenEncoderLarge.java

```diff
@@ -18,35 +18,33 @@ static int calculateTokensLarge(TokenEncoder tokenEncoder, int maxTokenCount, bo
 
         TreeMap<Integer, LinkedHashMap<Integer, RankNode>> rankMap = new TreeMap<>();
 
-        RankNode head = null;
-        RankNode prevNode = null;
+        RankNode head = null, prevNode = null;
         for (int i = 0; i < length + 1; i++) {
-            int encoded = tokenEncoder.encode(match, i, i + 2);
-            RankNode node = new RankNode(encoded, i);
-            if (head == null) {
-                head = node;
-            } else {
+            int rank = tokenEncoder.encode(match, i, i + 2);
+            RankNode node = new RankNode(rank, i, prevNode);
+            if (prevNode != null) {
                 prevNode.next = node;
-                node.prev = prevNode;
+            } else {
+                head = node;
             }
             prevNode = node;
 
-            rankMap.computeIfAbsent(encoded, k -> new LinkedHashMap<>()).put(i, node);
+            rankMap.computeIfAbsent(rank, k -> new LinkedHashMap<>()).put(i, node);
         }
         assert rankMap.containsKey(MAX_RANK);
-        while (rankMap.size() > 1) {
+        while (length > 2 && rankMap.size() > 1) {
             for (Iterator<RankNode> it = rankMap.pollFirstEntry().getValue().values().iterator(); it.hasNext(); ) {
                 RankNode minNode = it.next();
                 int minRank = minNode.rank;
                 assert minRank != MAX_RANK;
 
-                RankNode previousNode = minNode.prev;
-                RankNode nextNode = minNode.next;
-                RankNode nextNextNode = nextNode != null ? nextNode.next : null;
-                RankNode nextNextNextNode = nextNextNode != null ? nextNextNode.next : null;
+                RankNode previousNode = minNode.prev,
+                    nextNode = minNode.next,
+                    nextNextNode = nextNode.next,
+                    nextNextNextNode = nextNextNode.next;
```

Comment:
only nextNextNextNode can actually be null, since the last two are MAX_RANK values